chia/full_node/full_node.py

```diff
@@ -825,13 +858,41 @@ def get_peers_with_peak(self, peak_hash: bytes32) -> List:
         peers_with_peak: List = [c for c in self.server.all_connections.values() if c.peer_node_id in peer_ids]
         return peers_with_peak
 
+    async def update_wallets(
+        self, height: uint32, fork_height: uint32, peak_hash: bytes32, state_update: List[CoinRecord]
```

Comment:
I really think `state_update` should be a list of coin names, rather than full `CoinRecord`s. Right now we're doing database lookups for **every** coin being removed (twice, once to validate it and then once in `CoinStore`) just to be able to pass it in here. In this function we really only care about a few coins (most likely a small subset). If we would defer looking up the coin records, we would presumably save a lot of database I/O.

Reply:
an even more efficient way of doing it would be to have the protocol indicate coins as being spent (and maybe at what height) and just pass on the coin IDs to the wallet. Presumably the wallet will be able to look up its coins by ID and set the spent-bit itself.

Reply:
I don't think I added second lookup, it was there when _set_spent is called, I just returned the already looked up value.

Reply:
But I agree that can probably be optimized so there is only one lookup

Reply:
This is the object that wallet receives
```@dataclass(frozen=True)
@streamable
class CoinState(Streamable):
    coin: Coin
    spent_height: Optional[uint32]
    created_height: Optional[uint32]
```

Reply:
I have a PR where I'm eliminating that lookup, that's why I'm pointing it out. I think it would make sense to make this lookup only for the coins there are subscriptions on.

Reply:
Subscriptions can be for both puzzle_hash and coin_id. If we return just a coin id from this we'll miss spend updates for puzzle_hashes.